libtenzir/builtins/operators/sample.cpp

```diff
@@ -0,0 +1,214 @@
+  auto name() const -> std::string override {
+    return "tql2.sample";
+  }
+
+  auto location() const -> operator_location override {
+    // TODO: Figure this out
+    return operator_location::local;
+  }
+
+  auto
+  optimize(expression const&, event_order) const -> optimize_result override {
+    // TODO: Figure this out
+    return do_not_optimize(*this);
+  }
+
+  friend auto inspect(auto& f, sample_operator& x) -> bool {
+    return f.object(x)
+      .pretty_name("tenzir.plugins.sample.sample_operator")
+      // FIXME: Add mode_
+      .fields(f.field("delta", x.delta_));
+  }
+
+private:
+  mode mode_;
+  duration delta_;
+};
+
+class plugin final : public virtual operator_plugin<sample_operator>,
+                     public virtual operator_factory_plugin {
+public:
+  auto signature() const -> operator_signature override {
+    return {.transformation = true};
+  }
+
+  auto
+  make(invocation inv, session ctx) const -> failure_or<operator_ptr> override {
+    auto mode = located<std::string>{};
+    auto delta = located<duration>{};
+    // TODO: Decide name
+    TRY(argument_parser2::operator_("sample")
+          .add("mode", mode)
+          .add("delta", delta)
+          .parse(inv, ctx));
+    if (delta.inner.count() < 0) {
+      diagnostic::error("`delta` must be positive").primary(delta).emit(ctx);
+      return failure::promise();
+    }
+    auto mode_ = [&]() {
+      if (mode.inner == "log") {
+        return mode::log;
+      }
+      if (mode.inner == "log2") {
+        return mode::log2;
+      }
+      if (mode.inner == "log10") {
+        return mode::log10;
+      }
+      if (mode.inner == "sqrt") {
+        return mode::sqrt;
+      }
+      return mode::none;
+    }();
+    if (mode_ == mode::none) {
+      diagnostic::error(
+        R"(`mode` must be one of `"ln"`, "log2"`, `"log10"` or `"sqrt"`)")
+        .primary(mode)
+        .emit(ctx);
+      return failure::promise();
+    }
+    return std::make_unique<sample_operator>(mode_, delta.inner);
+  }
+
+  auto parse_operator(parser_interface& p) const -> operator_ptr override {
+    auto mode = located<std::string>{};
+    auto delta = located<duration>{};
+    auto parser
+      = argument_parser{"sample", "https://docs.tenzir.com/operators/sample"};
+    parser.add("--mode", mode, "<mode>");
+    parser.add("--delta", mode, "<delta>");
+    parser.parse(p);
+    if (delta.inner.count() < 0) {
+      diagnostic::error("`delta` must be positive").primary(delta).throw_();
+    }
+    auto mode_ = [&]() {
+      if (mode.inner == "log") {
+        return mode::log;
+      }
+      if (mode.inner == "log2") {
+        return mode::log2;
+      }
+      if (mode.inner == "log10") {
+        return mode::log10;
+      }
+      if (mode.inner == "sqrt") {
+        return mode::sqrt;
+      }
+      return mode::none;
+    }();
+    if (mode_ == mode::none) {
+      diagnostic::error(
+        R"(`mode` must be one of `"ln"`, "log2"`, `"log10"` or `"sqrt"`)")
```

Comment:
In other diagnostics we put the available options in the hint, like for example when you try to call a function that doesn't exist. Please use that style consistently.